File: games/nim/bots/ideal_bot.cpp
#include <cstdio>

const int MAX_COUNT = (int)1e5;
int heap_sizes[MAX_COUNT];

int main()
{
    while (1)
    {
        int heaps_count = 0;
        while (scanf("%d", &heap_sizes[heaps_count]) == 1)
        {
            ++heaps_count;
        }

        int global_xor = 0;
        for (int i = 0; i < heaps_count; ++i)
        {
            global_xor ^= heap_sizes[i];
        }

        if (global_xor)
        {
            for (int i = 0; i < heaps_count; ++i)
            {
                int new_size = global_xor ^ heap_sizes[i];
                if (heap_sizes[i] >= new_size)
                {
                    printf("%d %d\n", i, heap_sizes[i] - new_size);
                    break;
                }
            }
        }
        else
        {
            for (int i = 0; i < heaps_count; ++i)
            {
                if (heap_sizes[i])
                {
                    printf("%d %d\n", i, 1);
                    break;
                }
            }
        }

        fflush(stdout);
    }
    return 0;
}

File: games/tic-tac-toe/bots/ideal_bot.py
#!/usr/bin/env python3

import sys
import random
import operator
from functools import reduce


while True:
    heap_sizes = [int(n) for n in input().split(' ')]
    global_xor = reduce(operator.xor, heap_sizes)
    if global_xor != 0:
        for heap_number, size in enumerate(heap_sizes):
            new_size = global_xor ^ size
            if size > new_size:
                print(heap_number, size - new_size)
                break
    else:
        # Random move
        allowed = [i for i, size in enumerate(heap_sizes) if size > 0]
        heap_number = random.choice(allowed)
        removed_stones = random.randint(1, heap_sizes[heap_number])
        print(heap_number, removed_stones)
    sys.stdout.flush()
